src/interface/data_collection.hpp

```diff
@@ -0,0 +1,80 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_DATA_COLLECTION_HPP_
+#define INTERFACE_DATA_COLLECTION_HPP_
+
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace parthenon {
+class Mesh;
+
+/// The DataCollection class is an abstract container that contains at least a
+/// "base" container of some type (e.g., of MeshData or MeshBlockData) plus
+/// additional containers identified by string labels.
+/// Current usage includes (but is not limited to) storing MeshBlockData for different
+/// stages in multi-stage drivers or the corresponding MeshBlockPacks in a
+/// DataCollection of MeshData.
+template <typename T>
+class DataCollection {
+ public:
+  DataCollection() {
+    containers_["base"] = std::make_shared<T>(); // always add "base" container
+    pmy_mesh_ = nullptr;
+  }
+
+  void SetMeshPointer(Mesh *pmesh) { pmy_mesh_ = pmesh; }
+
+  std::shared_ptr<T> Add(const std::string &label, const std::shared_ptr<T> &src);
+  std::shared_ptr<T> Add(const std::string &label) {
+    // error check for duplicate names
+    auto it = containers_.find(label);
+    if (it != containers_.end()) {
+      return it->second;
+    }
+    containers_[label] = std::make_shared<T>();
+    return containers_[label];
+  }
+
+  std::shared_ptr<T> &Get() { return containers_["base"]; }
+  std::shared_ptr<T> &Get(const std::string &label) {
+    auto it = containers_.find(label);
+    if (it == containers_.end()) {
+      throw std::runtime_error("Container " + label + " does not exist in collection.");
+    }
+    return it->second;
+  }
+
+  std::shared_ptr<T> &GetOrAdd(const std::string &mbd_label, const int &partition_id);
```

Comment:
I feel like `GetOrAdd` could be unified with `Get/Add` a bit better. For example, perhaps there could be a version with signature
```C++
  std::shared_ptr<T> &GetOrAdd(const std::string &mbd_label, const std::string &base);
```
to replace separate `Get` and `Add` calls elsewhere in the code.

Also perhaps the signature for the one with an int partition id should have a default for the case of only 1 partition. e.g.,
```suggestion
  std::shared_ptr<T> &GetOrAdd(const std::string &mbd_label, const int partition_id = 0);
```
This can all be saved for another time (or never) though.

Reply:
The main reason why I didn't go a more "unified" version as the `GetOrAdd` is specific to the `DataCollection<MeshData>` and not implemented for the `DataCollection<MeshBlockData>` (both implementation would be distinct whereas the high level `GetOrAdd` work in their templated form.